xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,322 @@
+    """
+    name: str = "pdl.range"
+    arguments: Annotated[VarOperand, AnyPDLType | RangeType[AnyPDLType]]
+    result: Annotated[OpResult, RangeType[AnyPDLType]]
+
+    def verify_(self) -> None:
+
+        def get_type_or_elem_type(arg: SSAValue) -> Attribute:
+            if isinstance(arg.typ, RangeType):
+                arg_typ = cast(RangeType[AnyPDLType], arg.typ)
+                return arg_typ.element_type
+            else:
+                return arg.typ
+
+        if len(self.arguments) > 0:
+            elem_type = get_type_or_elem_type(self.result)
+
+            for arg in self.arguments:
+                if cur_elem_type := get_type_or_elem_type(arg) != elem_type:
+                    raise VerifyException(
+                        f"All arguments must have the same type or be an array  \
+                          of the corresponding element type. First element type:\
+                          {elem_type}, current element type: {cur_elem_type}")
+
+
+@irdl_op_definition
+class ReplaceOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlreplace-mlirpdlreplaceop
+    """
+    name: str = "pdl.replace"
+    opValue: Annotated[Operand, OperationType]
+    replOperation: Annotated[OptOperand, OperationType]
+    replValues: Annotated[VarOperand, ValueType | ArrayAttr[ValueType]]
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+
+@irdl_op_definition
+class ResultOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlresult-mlirpdlresultop
+    """
+    name: str = "pdl.result"
+    index: OpAttr[IntegerAttr[IntegerType]]
+    parent_: Annotated[Operand, OperationType]
+    val: Annotated[OpResult, ValueType]
+
+
+@irdl_op_definition
+class ResultsOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlresults-mlirpdlresultsop
+    """
+    name: str = "pdl.results"
+    index: OpAttr[IntegerAttr[IntegerType]]
+    parent_: Annotated[Operand, OperationType]
+    val: Annotated[OpResult, ValueType | ArrayAttr[ValueType]]
+
+
+@irdl_op_definition
+class RewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlrewrite-mlirpdlrewriteop
+    """
+    name: str = "pdl.rewrite"
+    root: Annotated[OptOperand, OperationType]
+    # name of external rewriter function
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OptOpAttr[StringAttr]
+    externalArgs: Annotated[VarOperand, AnyPDLType]
+    body: OptRegion
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+    def verify_(self) -> None:
+        if 'name' in self.attributes:
+            if not isinstance(self.attributes['name'], StringAttr):
+                raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, root: SSAValue | None, external_args: list[SSAValue],
+            body: Region | None) -> RewriteOp:
+
+        operands: list[SSAValue | list[SSAValue]] = []
+        if root is not None:
```

Comment:
```suggestion
        if root:
```

Reply:
Checking for `None` explicitly seems clearer. Who knows if someone might override `__bool__` or `__len__` on SSAValue in the future

Reply:
Okay, whatever you prefer :+1: